features/steps/steps.py

```diff
@@ -283,4 +329,30 @@ def step_impl(context, representation_id):
                     errors.append(representation_shape_error(inst, representation_id))
     
     handle_errors(context, errors)
-    
+
+@then("The {entity} must have no duplicate points")
+def step_impl(context, entity):
+    if getattr(context, 'applicable', True):
+        errors = []
+        precision = get_sending_system_precision(context)
+        for instance in context.instances:
+            points = get_entity_point_components(instance)
+            points_coordinates = [point.Coordinates for point in points]
+            points_coordinates_rounded = []
+            for point_coordinates in points_coordinates:
+                point_coordinates_rounded = tuple(round(point / precision) * precision for point in point_coordinates)
```

Comment:
This is clever because you can use sets, but it's not geometrically what is intended.

This is the unpleasant consequence of geometric tolerance. You loose the ability to certain things efficiently and loose desirable mathematical traits such as transitiveness (a = b, b = c ---> a = c).

I would propose to do a bruteforce approach using itertools.combinations(2) to take all 2-combinations from the list of points and then check whether the distance between those two points (2-norm, numpy.linalg.norm or just good old pythagoras) is within tolerance.

Reply:
Done :)